### PR TITLE
Fix building for Python 3.8

### PR DESCRIPTION
**Pull request by bryanwweber**
Thanks for contributing code! Please include a description of your change and
check your PR against the list below (for further questions, refer to the
[contributing guide](https://github.com/Cantera/cantera/blob/master/CONTRIBUTING.md)).

- [x] There is a clear use-case for this code change
- [x] The commit message has a short title & references relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage

**Changes proposed in this pull request**

- Bump the minimum Cython version when building for Python 3.8
- Disable deprecation warnings on macOS when compiling `_cantera.cpp` that are caused by a deprecated feature that Cython handles

### CODE REVIEW

**Comment by speth** [comment]
Do the compiler warnings go away with Cython 0.29.14? If so, can we just set that to the minimum version and not have to worry about disabling deprecation warnings?

**Comment by bryanwweber** [comment]
For reference, I'm using Python 3.8.0 on macOS and Linux built from the source tarball by pyenv, and on Windows Python 3.8.0 from the Anaconda distribution.

The warnings are present only on macOS and only with Cython 0.29.14. For reference, these are the warnings I get:

```
g++ -o build/temp-py/_cantera.os -c -std=c++0x -O3 -Wno-inline -g -fPIC -DNDEBUG -Iinclude -I/Users/bryan/.pyenv/versions/3.8.0/include/python3.8 -I/Users/bryan/.pyenv/versions/cantera-dev/lib/python3.8/site-packages/numpy/core/include -I/Users/bryan/miniconda3/envs/cantera-dev/include interfaces/cython/cantera/_cantera.cpp
... (NumPy warnings removed)
interfaces/cython/cantera/_cantera.cpp:105531:3: warning: 'tp_print' is deprecated [-Wdeprecated-declarations]
  0, /*tp_print*/
  ^
... (hundreds of messages differing only in the source line)
```

The relevant lines from `_cantera.cpp` are
```
  #if PY_VERSION_HEX >= 0x030800b4 && PY_VERSION_HEX < 0x03090000
  0, /*tp_print*/
  #endif
```
which is basically repeated for each `PyTypeObject` that is generated.

These warnings are not present on Windows or on Linux for Cython >=0.29.12. With Cython < 0.29.12, the compile of `_cantera.cpp` fails with (here from Windows, but the same basic reason about `__Pyx_PyCode_New` happens on at least macOS even if the specific message is different):
```
cl /Fobuild\temp-py\_cantera.obj /c interfaces\cython\cantera\_cantera.cpp /EHsc /MD /nologo /D_SCL_SECURE_NO_WARNINGS /D_CRT_SECURE_NO_WARNINGS /O2 /Zi /Fdbuild\temp-py\_cantera.obj.pdb /DNDEBUG /Iinclude /IC:\Users\bww09001\AppData\Local\Continuum\Miniconda3\envs\cantera-dev\include /IC:\Users\bww09001\AppData\Local\Continuum\Miniconda3\envs\cantera-dev\lib\site-packages\numpy\core\include /IC:\Users\bww09001\AppData\Local\Continuum\Miniconda3\envs\cantera-dev\Library\Include
_cantera.cpp
... (NumPy warnings removed)
interfaces\cython\cantera\_cantera.cpp(116964): warning C4003: not enough arguments for function-like macro invocation '__Pyx_PyCode_New'
interfaces\cython\cantera\_cantera.cpp(116964): error C2059: syntax error: ')'
```

Interestingly, Cython 0.29.12 on Windows and 0.29.13 on macOS do not generate the warnings that https://github.com/cython/cython/pull/3171 and https://github.com/cython/cython/pull/3201 were merged to fix. In fact, they don't generate any warnings at all (except the standard NumPy one). 🤷‍♂

Basically, I have no idea what's going on here. The changes in this PR allow building with Python 3.8.0 without spurious warnings. That's all I know 😅

File: SConstruct [hunk 1]
```diff
@@ -1293,6 +1294,11 @@ if env['python_package'] != 'none':
                 "Found {0} but {1} or newer is required".format(
                 python_version, python_min_version))
             warn_no_full_package = True
+        elif python_version >= LooseVersion("3.8"):
+            # Reset the minimum Cython version if the Python version is 3.8 or higher
+            # Due to internal changes in the CPython API, more recent versions of
+            # Cython are necessary to build for Python 3.8
+            cython_min_version = LooseVersion("0.29.12")
```
Comment (speth) [inline SConstruct]: I think it would be helpful for the comments here to be slightly more clear that there's nothing Cantera-specific about any of this: Only Cython >= 0.29.12 is compatible with Python 3.8, and we're just setting this dependency so that users get a more useful error message than the compilation errors that would otherwise result.

File: interfaces/cython/SConscript [hunk 2]
```diff
@@ -47,6 +47,13 @@ localenv.Prepend(CPPPATH=[Dir('#include'), inc, numpy_include])
 localenv.Prepend(LIBS=localenv['cantera_libs'])
 if localenv['OS'] == 'Darwin':
     localenv.Append(LINKFLAGS='-undefined dynamic_lookup')
+    # Don't print deprecation warnings for internal Python changes.
+    # It seems that only clang on macOS prints these deprecation warnings.
+    # Only applies to Python 3.8 and Cython < 3.0. The source of the warnings
+    # will be removed after Python 3.8, so the minimum version of Cython will
+    # need to change for Python 3.9.
+    if LooseVersion(py_version) == LooseVersion("3.8"):
```
Comment (speth) [inline interfaces/cython/SConscript]: Given the documentation for `distutils.sysconfig.get_python_version()`, this could just be:
```suggestion
    if py_version == "3.8":
```

**Review by speth — APPROVED** [review]
Just a couple of very minor suggestions. Feel free to address as you like.

**Comment by bryanwweber** [comment]
For what it's worth, there were various other internal changes in CPython in the beta period for 3.8, such that the Cython maintainer states that Cython 0.29.12 is the first version compatible with the released CPython 3.8: https://bugs.python.org/issue37221#msg347468

Also related, with more information for future researchers: https://github.com/python/cpython/pull/14193 and https://bugs.python.org/issue37250 and https://github.com/cython/cython/issues/2976
